File: src/model/ct2.py
from __future__ import annotations

import io
import struct
from dataclasses import dataclass
from enum import Enum
from pathlib import Path
from typing import BinaryIO, Iterator

import numpy as np


class Ct2DataType(Enum):
    FLOAT32 = 0
    INT8 = 1
    INT16 = 2
    INT32 = 3
    FLOAT16 = 4
    BFLOAT16 = 5

    @classmethod
    def from_id(cls, dtype_id: int) -> "Ct2DataType":
        try:
            return cls(dtype_id)
        except ValueError as exc:
            raise ValueError(f"unsupported CTranslate2 dtype id {dtype_id}") from exc

    @property
    def label(self) -> str:
        return {
            Ct2DataType.FLOAT32: "float32",
            Ct2DataType.INT8: "int8",
            Ct2DataType.INT16: "int16",
            Ct2DataType.INT32: "int32",
            Ct2DataType.FLOAT16: "float16",
            Ct2DataType.BFLOAT16: "bfloat16",
        }[self]


@dataclass(slots=True, frozen=True)
class TensorInfo:
    name: str
    shape: list[int]
    dtype: Ct2DataType
    offset: int
    nbytes: int

    @property
    def element_count(self) -> int:
        count = 1
        for dim in self.shape:
            count *= dim
        return count


@dataclass(slots=True, frozen=True)
class TensorDataF32:
    info: TensorInfo
    values: list[float]


class Ct2ModelBin:
    def __init__(
        self,
        path: Path,
        *,
        version: int,
        spec_name: str,
        revision: int,
        tensors: dict[str, TensorInfo],
    ) -> None:
        self.path = path
        self.version = version
        self.spec_name = spec_name
        self.revision = revision
        self._tensors = tensors

    @classmethod
    def open(cls, path: str | Path) -> "Ct2ModelBin":
        path = Path(path)
        with path.open("rb") as handle:
            version = _read_u32(handle)
            spec_name = _read_string(handle)
            revision = _read_u32(handle)
            tensor_count = _read_u32(handle)
            tensors: dict[str, TensorInfo] = {}

            for _ in range(tensor_count):
                name = _read_string(handle)
                rank = _read_u8(handle)
                shape = [_read_u32(handle) for _ in range(rank)]
                dtype = Ct2DataType.from_id(_read_u8(handle))
                nbytes = _read_u32(handle)
                offset = handle.tell()
                handle.seek(nbytes, io.SEEK_CUR)

                tensors[name] = TensorInfo(
                    name=name,
                    shape=shape,
                    dtype=dtype,
                    offset=offset,
                    nbytes=nbytes,
                )

            alias_count = _read_u32(handle)
            for _ in range(alias_count):
                _read_string(handle)
                _read_string(handle)

        return cls(
            path=path,
            version=version,
            spec_name=spec_name,
            revision=revision,
            tensors=tensors,
        )

    def tensor(self, name: str) -> TensorInfo:
        try:
            return self._tensors[name]
        except KeyError as exc:
            raise KeyError(f"tensor not found in model.bin: {name}") from exc

    def read_tensor_f32(self, name: str) -> TensorDataF32:
        info = self.tensor(name)
        with self.path.open("rb") as handle:
            handle.seek(info.offset)
            raw = handle.read(info.nbytes)

        if len(raw) != info.nbytes:
            raise ValueError(
                f"failed to read tensor {name}: expected {info.nbytes} bytes, got {len(raw)}"
            )

        if info.dtype is Ct2DataType.FLOAT32:
            values = np.frombuffer(raw, dtype="<f4").astype(np.float32, copy=False)
            result = values.tolist()
        elif info.dtype is Ct2DataType.FLOAT16:
            values = np.frombuffer(raw, dtype="<f2").astype(np.float32)
            result = values.tolist()
        else:
            raise ValueError(
                f"tensor {name} has dtype {info.dtype.label}, not a float tensor"
            )

        if len(result) != info.element_count:
            raise ValueError(
                f"tensor {info.name} element count mismatch: "
                f"expected {info.element_count}, got {len(result)}"
            )

        return TensorDataF32(info=info, values=result)

    def tensor_names(self) -> Iterator[str]:
        return iter(self._tensors.keys())


def _read_exact(handle: BinaryIO, size: int) -> bytes:
    data = handle.read(size)
    if len(data) != size:
        raise ValueError(f"unexpected EOF while reading {size} bytes")
    return data


def _read_u8(handle: BinaryIO) -> int:
    return struct.unpack("<B", _read_exact(handle, 1))[0]


def _read_u16(handle: BinaryIO) -> int:
    return struct.unpack("<H", _read_exact(handle, 2))[0]


def _read_u32(handle: BinaryIO) -> int:
    return struct.unpack("<I", _read_exact(handle, 4))[0]


def _read_string(handle: BinaryIO) -> str:
    byte_len = _read_u16(handle)
    if byte_len == 0:
        raise ValueError("invalid zero-length string in model.bin")

    raw = _read_exact(handle, byte_len)
    if raw[-1] != 0:
        raise ValueError("model.bin string is missing trailing NUL")

    try:
        return raw[:-1].decode("utf-8")
    except UnicodeDecodeError as exc:
        raise ValueError("model.bin contained invalid UTF-8") from exc

File: src/model/ct2.rs
use std::collections::BTreeMap;
use std::fs::File;
use std::io::{BufReader, Read, Seek, SeekFrom};
use std::path::{Path, PathBuf};

use anyhow::{Context, Result, bail};
use half::f16;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Ct2DataType {
    Float32,
    Int8,
    Int16,
    Int32,
    Float16,
    BFloat16,
}

impl Ct2DataType {
    fn from_id(id: u8) -> Result<Self> {
        match id {
            0 => Ok(Self::Float32),
            1 => Ok(Self::Int8),
            2 => Ok(Self::Int16),
            3 => Ok(Self::Int32),
            4 => Ok(Self::Float16),
            5 => Ok(Self::BFloat16),
            _ => bail!("unsupported CTranslate2 dtype id {id}"),
        }
    }

    pub fn label(self) -> &'static str {
        match self {
            Self::Float32 => "float32",
            Self::Int8 => "int8",
            Self::Int16 => "int16",
            Self::Int32 => "int32",
            Self::Float16 => "float16",
            Self::BFloat16 => "bfloat16",
        }
    }
}

#[derive(Debug, Clone)]
pub struct TensorInfo {
    pub name: String,
    pub shape: Vec<usize>,
    pub dtype: Ct2DataType,
    pub offset: u64,
    pub nbytes: usize,
}

impl TensorInfo {
    pub fn element_count(&self) -> usize {
        self.shape.iter().product()
    }
}

#[derive(Debug)]
pub struct Ct2ModelBin {
    path: PathBuf,
    pub version: u32,
    pub spec_name: String,
    pub revision: u32,
    tensors: BTreeMap<String, TensorInfo>,
}

#[derive(Debug, Clone)]
pub struct TensorDataF32 {
    pub info: TensorInfo,
    pub values: Vec<f32>,
}

impl Ct2ModelBin {
    pub fn open(path: impl AsRef<Path>) -> Result<Self> {
        let path = path.as_ref().to_path_buf();
        let mut reader = BufReader::new(
            File::open(&path).with_context(|| format!("failed to open {}", path.display()))?,
        );

        let version = read_u32(&mut reader)?;
        let spec_name = read_string(&mut reader)?;
        let revision = read_u32(&mut reader)?;
        let tensor_count = read_u32(&mut reader)? as usize;
        let mut tensors = BTreeMap::new();

        for _ in 0..tensor_count {
            let name = read_string(&mut reader)?;
            let rank = read_u8(&mut reader)? as usize;
            let mut shape = Vec::with_capacity(rank);
            for _ in 0..rank {
                shape.push(read_u32(&mut reader)? as usize);
            }
            let dtype = Ct2DataType::from_id(read_u8(&mut reader)?)?;
            let nbytes = read_u32(&mut reader)? as usize;
            let offset = reader.stream_position()?;
            reader.seek(SeekFrom::Current(nbytes as i64))?;
            tensors.insert(
                name.clone(),
                TensorInfo {
                    name,
                    shape,
                    dtype,
                    offset,
                    nbytes,
                },
            );
        }

        let alias_count = read_u32(&mut reader)? as usize;
        for _ in 0..alias_count {
            let _ = read_string(&mut reader)?;
            let _ = read_string(&mut reader)?;
        }

        Ok(Self {
            path,
            version,
            spec_name,
            revision,
            tensors,
        })
    }

    pub fn tensor(&self, name: &str) -> Result<&TensorInfo> {
        self.tensors
            .get(name)
            .with_context(|| format!("tensor not found in model.bin: {name}"))
    }

    pub fn read_tensor_f32(&self, name: &str) -> Result<TensorDataF32> {
        let info = self.tensor(name)?.clone();
        let mut reader = BufReader::new(
            File::open(&self.path)
                .with_context(|| format!("failed to reopen {}", self.path.display()))?,
        );
        reader.seek(SeekFrom::Start(info.offset))?;
        let mut bytes = vec![0_u8; info.nbytes];
        reader.read_exact(&mut bytes)?;

        let values: Vec<f32> = match info.dtype {
            Ct2DataType::Float32 => bytes
                .chunks_exact(4)
                .map(|chunk| f32::from_le_bytes([chunk[0], chunk[1], chunk[2], chunk[3]]))
                .collect(),
            Ct2DataType::Float16 => bytes
                .chunks_exact(2)
                .map(|chunk| {
                    let bits = u16::from_le_bytes([chunk[0], chunk[1]]);
                    f16::from_bits(bits).to_f32()
                })
                .collect(),
            _ => bail!(
                "tensor {name} has dtype {}, not a float tensor",
                info.dtype.label()
            ),
        };

        if values.len() != info.element_count() {
            bail!(
                "tensor {} element count mismatch: expected {}, got {}",
                info.name,
                info.element_count(),
                values.len()
            );
        }

        Ok(TensorDataF32 { info, values })
    }

    pub fn tensor_names(&self) -> impl Iterator<Item = &str> {
        self.tensors.keys().map(String::as_str)
    }
}

fn read_u8(reader: &mut BufReader<File>) -> Result<u8> {
    let mut buf = [0_u8; 1];
    reader.read_exact(&mut buf)?;
    Ok(buf[0])
}

fn read_u16(reader: &mut BufReader<File>) -> Result<u16> {
    let mut buf = [0_u8; 2];
    reader.read_exact(&mut buf)?;
    Ok(u16::from_le_bytes(buf))
}

fn read_u32(reader: &mut BufReader<File>) -> Result<u32> {
    let mut buf = [0_u8; 4];
    reader.read_exact(&mut buf)?;
    Ok(u32::from_le_bytes(buf))
}

fn read_string(reader: &mut BufReader<File>) -> Result<String> {
    let byte_len = read_u16(reader)? as usize;
    if byte_len == 0 {
        bail!("invalid zero-length string in model.bin");
    }
    let mut buf = vec![0_u8; byte_len];
    reader.read_exact(&mut buf)?;
    if buf.last() != Some(&0) {
        bail!("model.bin string is missing trailing NUL");
    }
    String::from_utf8(buf[..buf.len() - 1].to_vec())
        .with_context(|| "model.bin contained invalid UTF-8")
}
